[Bug 15588] Script editor performance improvements

peter-b opened the pull request:
Improve script editor performance when working with huge scripts in a number of ways:
- Only reset the script editor field styles (font style / size) if they don't match the current preferences setting
- Align images in the script editor's handler list more efficiently (thanks to @runrevmark for the fix)
- Only resize script editor fields when necessary
- Cache the whole main script editor field to avoid re-populating when switching repeatedly between a small set of tabs

This is for @trevordevore :wink:

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/livecode/livecode-ide/794)

<!-- Reviewable:end -->

peter-b committed:
[Perf][Bug 15588] Only update script editor styles if preferences don't match.

Users don't change their script editor display preferences very often,
and it's very slow to reset all the fonts of the script editor due to
need to completely re-layout the field.  For large scripts, only
resetting the editor field's styles when they don't match the current
preferences can save a significant amount of time.  This patch speeds
up switching tabs in the script editor by approximately 30%.

For example, on my system, switching to the revidelibrary script was
about 1.9 seconds and this patch reduces the time to about 1.1 s.

peter-b committed:
[Perf][Bug 15588] Align script editor handler list text more efficiently

Looping over the lines of the field and adjusting the text alignment
is quite slow -- instead, build the alignment information into the
generated htmlText directly.  This saves about 2/3 of the time needed
to update the left bar's handler list (0.4 seconds out of 0.6 seconds
when opening the IDE library script on my machine).

peter-b committed:
[Perf][Bug 15588] Optimise editor main pane resizing.

Save about 0.4 seconds on opening or switching to a large script by
optimizing the relayout process.  Only resize each control once per
relayout, and only if the desired bounds are different from the
current bounds.

peter-b committed:
[Perf][Bug 15588] Cache script editor fields for object scripts

For very large scripts, resetting and reformatting the script field is
very costly -- sometimes over a second.  This is work that's largely
duplicative, especially when switching frequently between editor tabs
(which can occur because you have a script library open and you are
referring to its source code while writing a handler that makes use of
it).

By keeping editor fields for recent tabs around, invisibly, and
swapping them in and out as the user switches tabs, a substantial
speed up can be obtained when working with large scripts.

This patch hacks something into the script editor to achieve this.
Fortunately, it's relatively self-contained.  It currently caches a
maximum of 10 editor fields per script editor window, and makes the
field update cost of switching tabs negligible when there are no
script or layout changes required.

peter-b committed:
[Bug 15588] Add release note.

trevordevore commented:
You're a good man @peter-b :-)

peter-b commented:
@trevordevore On my Linux VM, switching between two 10kloc scripts is down to <1 s from ~4 sec.  Hope that's good enough for the time being...?

trevordevore commented:
@peter-b That sounds fantastic!

livecodeali commented:
_[Toolset/palettes/script editor/behaviors/revseeditorbehavior.livecodescript, line 186 [r1]](https://reviewable.io:443/reviews/livecode/livecode-ide/794#-K8R2uzDqGYKc0eZM3_T:-K8R2uzDqGYKc0eZM3_U:-681299376) ([raw file](https://github.com/livecode/livecode-ide/blob/5aba4a08537f0a20acb0e97e7097894092ebd471/Toolset/palettes/script editor/behaviors/revseeditorbehavior.livecodescript#L186)):_
I assume you meant to use tFont and tFontSize here rather than get them again?

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode-ide/794)_

<!-- Sent from Reviewable.io -->

livecodeali commented:
_[Toolset/palettes/script editor/behaviors/revseeditorbehavior.livecodescript, line 197 [r1]](https://reviewable.io:443/reviews/livecode/livecode-ide/794#-K8R3YTGAF7RmFoUVE4G:-K8R3YTHOOe9FImrZcCz:-897005896) ([raw file](https://github.com/livecode/livecode-ide/blob/5aba4a08537f0a20acb0e97e7097894092ebd471/Toolset/palettes/script editor/behaviors/revseeditorbehavior.livecodescript#L197)):_
Should be tFontSize (and again below)

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode-ide/794)_

<!-- Sent from Reviewable.io -->

livecodeali commented:
_[Toolset/palettes/script editor/behaviors/revseeditorbehavior.livecodescript, line 432 [r1]](https://reviewable.io:443/reviews/livecode/livecode-ide/794#-K8R50BbO7EQF_fgQJRd:-K8R50Bcsk3cE9FOTbOB:509055492) ([raw file](https://github.com/livecode/livecode-ide/blob/5aba4a08537f0a20acb0e97e7097894092ebd471/Toolset/palettes/script editor/behaviors/revseeditorbehavior.livecodescript#L432)):_
Might be worth putting seGetRuggedId(pObjectID) into a variable to avoid calling it twice

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode-ide/794)_

<!-- Sent from Reviewable.io -->

livecodeali commented:
_[Toolset/palettes/script editor/behaviors/revseeditorbehavior.livecodescript, line 5281 [r1]](https://reviewable.io:443/reviews/livecode/livecode-ide/794#-K8R6f-eEVr2cR1Yqs31:-K8R6f-eEVr2cR1Yqs32:311730152) ([raw file](https://github.com/livecode/livecode-ide/blob/5aba4a08537f0a20acb0e97e7097894092ebd471/Toolset/palettes/script editor/behaviors/revseeditorbehavior.livecodescript#L5281)):_
I _think_ it's currently ok to remove array elements whilst iterating isn't it? But not when adding elements? IIRC we don't rehash when removing. But it might be better to iterate over the lines of the keys instead so as not to give the impression this is something we encourage... your call though :-)

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode-ide/794)_

<!-- Sent from Reviewable.io -->

livecodeali commented:
_[Toolset/palettes/script editor/behaviors/revseeditorbehavior.livecodescript, line 5322 [r1]](https://reviewable.io:443/reviews/livecode/livecode-ide/794#-K8R9MQ0dynkBt3I0kz_:-K8R9MQ0dynkBt3I0kza:-1853840787) ([raw file](https://github.com/livecode/livecode-ide/blob/5aba4a08537f0a20acb0e97e7097894092ebd471/Toolset/palettes/script editor/behaviors/revseeditorbehavior.livecodescript#L5322)):_
EditorFieldCacheSwap and __EditorFieldCacheFlush need lock/unlock messages as they are creating and deleting objects and changing names - that will trigger all sorts of IDE updates that might cause focus / active stack changes.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode-ide/794)_

<!-- Sent from Reviewable.io -->

livecodeali commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode-ide/794)_

<!-- Sent from Reviewable.io -->

peter-b committed:
[Bug 15588] Update editor text styles correctly.

Fix some issues introduced in commit ae03b2f, where cached style
preferences were not used correctly.

peter-b committed:
[Bug 15588] Lock messages while performing editor field cache operations

Avoid IDE updates and focus or active stack changes when switching
script editor tabs.

peter-b committed:
[Bug 15588] Slightly optimise "rugged ID" use in script editor

Avoid calling `seGetRuggedId()` repeatedly for the same object.

peter-b commented:
@livecodeali Pull request updated!

livecodeali commented:
@livecode-vulcan review ok 4162031

livecode-vulcan commented:
💙 review by @livecodeali ok 416203124f6be630a14a8d67fe3726f291049992
